### ISSUE TITLE
make -j4 MACHINE=kvm_x86_64 all whilst in the build-env container fails

### ISSUE THREAD

**Issue by fire-ant** [opencomputeproject/onie#959]
I am folding the instructions letter for letter @ https://github.com/opencomputeproject/onie/tree/master/contrib/build-env#readme.

Whilst I have no problem getting the example build to run with `make -j4 MACHINE=kvm_x86_64 all` I dont see any images subsequently output in the ${HOME}/src/onie/build/images. 

I can upload logs if this directory is still actively maintained or just take a hint if it was broken long ago and I can't find a clue to that today.

Alternatively, if anyone has a way to build a kvm image from scratch Id also welcome that

**Comment by ehdoyle** [comment]
Hi - do post the logs, the checkout, the invocation, etc.
 If you can re run the build  with a -j1 that might narrow down where the problem is. Frequently the source of the problem gets buried in parallel builds.
Given that this sounds like a build environment config problem, I'd suggest using  [Dedicated User Environment](https://github.com/CumulusNetworks/DUE)  to create a Debian 9 Docker container build environment build ( granted, this is not well advertised yet, but has been how release builds have been generated.)
[https://github.com/CumulusNetworks/DUE/tree/master/templates/onie](https://github.com/CumulusNetworks/DUE/tree/master/templates/onie)
That will produce a standard build environment that can be used as a baseline for debugging any subsequent issues.
There's a tutorial video on this here: [YouTube](https://www.youtube.com/watch?v=-5onRbZA0QQ) along with the documentation on DUE

**Comment by fire-ant** [comment]
Alex, highly appreciated :) I tried 

`due --run --build --jobs 4 --machine kvm_x86_64 --build-targets all`

and it appeared that it was breaking around an upper-case char in a path that Crosstool-NG did not enjoy at all. And I dug deep enough to find everything you have recommended at breakfast and currently from within a Debian-9 DUE container, build  passes over (where I think it previously broke): 

`line 1408: CT_DoKernelTupleValues: command not found
[ERROR]  
[ERROR]  >>
[ERROR]  >>  Build failed in step '(top-level)'
[ERROR]  >>
[ERROR]  >>  Error happened in: CT_DoBuildTargetTuple[scripts/functions@1408]
[ERROR]  >>        called from: CT_LoadConfig[scripts/functions@31]
[ERROR]  >>        called from: main[scripts/crosstool-NG.sh@22]
[ERROR]  >>
[ERROR]  >>  For more info on this error, look at the file: ''
[ERROR]  >>  There is a list of known issues, some with workarounds, in:
[ERROR]  >>      '/workspaces/weave/cx-customer-dt-schiff/libvirt-lab/onie/build/crosstool-ng/crosstool-ng-1.23.0/docs/B - Known issues.txt'
[ERROR]  >>
[ERROR]  >>  If you feel this is a bug in crosstool-NG, report it at:
[ERROR]  >>      https://github.com/crosstool-ng/crosstool-ng/issues/
[ERROR]  >>
[ERROR]  >>  Make sure your report includes all the information pertinent to this issue.
[ERROR]  >>  Read the bug reporting guidelines here:
[ERROR]  >>      http://crosstool-ng.github.io/support/
[ERROR]  
[ERROR]  (elapsed: 27013706:23.03)`

but it is happily not halting here. Ill revert with anything I find that breaks the run.

**Comment by fire-ant** [comment]
Hi Alex,  when running:

` ______________________________________________________________________
|
| /usr/local/bin/duebuild
| Invoked with:      --jobs 1 --machine kvm_x86_64 --build-targets all
|
| Build dir            [ ommitted]
| Build attempts       [ 1 ]
| Build jobs           [ 1 ]
| Build start at       [ Wed May 12 12:53:58 UTC 2021 ]
| Machine              [ kvm_x86_64 ]
| Build targets        [ all ]
|_____________________________________________________________________`

I get:

`line 1408: CT_DoKernelTupleValues: command not found
[ERROR]  
[ERROR]  >>
[ERROR]  >>  Build failed in step '(top-level)'
[ERROR]  >>
[ERROR]  >>  Error happened in: CT_DoBuildTargetTuple[scripts/functions@1408]
[ERROR]  >>        called from: CT_LoadConfig[scripts/functions@31]
[ERROR]  >>        called from: main[scripts/crosstool-NG.sh@22]
[ERROR]  >>
[ERROR]  >>  For more info on this error, look at the file: ''
[ERROR]  >>  There is a list of known issues, some with workarounds, in:
[ERROR]  >>      '/workspaces/weave/cx-customer-dt-schiff/libvirt-lab/onie/build/crosstool-ng/crosstool-ng-1.23.0/docs/B - Known issues.txt'
[ERROR]  >>
[ERROR]  >>  If you feel this is a bug in crosstool-NG, report it at:
[ERROR]  >>      https://github.com/crosstool-ng/crosstool-ng/issues/
[ERROR]  >>
[ERROR]  >>  Make sure your report includes all the information pertinent to this issue.
[ERROR]  >>  Read the bug reporting guidelines here:
[ERROR]  >>      http://crosstool-ng.github.io/support/
[ERROR]  
[ERROR]  (elapsed: 27013757:17.04)`

is that enough to discuss from? Im following DUE to a T and confident the env is clean.

**Comment by ehdoyle** [comment]
Hi -  a couple of thoughts
 - Make sure you've got a clean slate for building - either a fresh checkout, or a 'make distclean', perhaps followed up with a git status to detect any stray files. I've seen situations where make thinks some components built when they really didn't after a compile fail ( and, of course, these things are obnoxious to debug because the process of determining the problem tends to eliminate it...)
 - Is the directory you're building in NFS mounted, or part of the system doing the build? I've seen what I can only describe as 'weird build errors' with Docker accessing mounted filesystems.
 - Is DUE being run (and the Debian 9 container built ) from a source checkout, or are you using a .deb install ( latest official release [here](https://github.com/CumulusNetworks/DUE/releases/tag/v2.3.0)
 - I just tried building from within the container, using 
 due --run <select container>
 cd to ONIE build-config directory
 make -j1  MACHINE=kvm_x86_64 all demo
 ..and that built. 
-  Re-reading the first of the last two posts, did you have a case where it worked and one where it doesn't?
 To build outside of the container, try:
 cd onie/build-config
 due --run-image due-onie-build-debian-9 --command "make -j4  MACHINE=kvm_x86_64 all demo"
 
 That will use your current directory and try to run the make command in it. Could be I'm getting too clever somewhere with duebuild...

**Comment by fire-ant** [comment]
Hi - So I think one key part Im not mentioning is that Im using a VsCode dev container (https://code.visualstudio.com/docs/remote/containers) in which Im building DUE from source, which is pointed back at my docker host (on macOS). Initially this was doing some docker-in-docker things to try and keep the environment isolated but as the build is in a container Ive since pointed it back at the docker host itself. This introduces a few unforeseens that haven't been accounted for in DUE (and im not  surprised) as the use-case is incredibly nuanced. I think we can put this in the 'weird build errors' box under that premise. So, short of spinning up a nested VM and starting over I wouldn't mind a hint at whether you think DUE should be able to work with on Docker for Desktop, if its built/containerised with the right config.

**Comment by ehdoyle** [comment]
Well, I'd say it's an up-till-now-untested environment :).
 Most of the DUE use has been on Debian and some on Red hat, either on real hardware or in virtual machines. You might be able to get around filesystem issues by logging into the container and checking out ONIE in an area that isn't mounted (say, /tmp ) or a docker volume? That might eliminate the issues around mounting a local filesystem, although you'd have to watch out for your build area vanishing (/tmp) or make sure you access the docker volume correctly.  
 I do know that a VM using Docker should work - I did build testing with VirtualBox - as long as that VM isn't mounting a build area on the host fileystem (which is where I ran into this issue in the first place, trying to use a VM that didn't have enough local storage.)...for what that's worth?
